### DISCUSSION TITLE
How are relative references in the CSDL declared together with the xml:base

### DISCUSSION THREAD

**Discussion by elize-vdr** [OData/odata.net#2628] [Q&A]
I am parsing OData CSDL contract documents using this library. I am struggling to understand exactly how a refrence, let us say for now I am referring to OData V4, but this is applicable to V3 too, so referring to the OData CSDL standard the datandard for V4 states the following regarding references:
   
*The [edmx:Reference](http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part3-csdl/odata-v4.0-errata03-os-part3-csdl-complete.html#_Element_edmx:Reference) element MUST specify a Uri attribute. The Uri attribute uniquely identifies a model, so two references MUST NOT specify the same URI. The value of the Uri attribute SHOULD be a URL that locates a CSDL document describing the referenced model. If the URI is not dereferencable it SHOULD identify a well-known schema. The value of the Uri attribute MAY be an absolute or relative URI; relative URIs are relative to the xml:base attribute, see [[XML‑Base]](http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part3-csdl/odata-v4.0-errata03-os-part3-csdl-complete.html#BMXMLBase).*

the "XML-Base" W3 standard at that link states that:

*The attribute xml:base [may](https://www.w3.org/TR/xmlbase/#dt-must) be inserted in XML documents to specify a base URI other than the base URI of the document or external entity.*

So from the above if there is a relative URI being used for an <edmx:Reference>, for example:
```
<emdx: Reference <edmx:Reference Uri="./CORE_AP242/CsdlSchema.xml">
	<edmx:Include Alias="CORE_AP242" Namespace="CORE_APA242_SCHEMA_V_1_0"/>
</edmx:Reference>
```
So this is what is unclear to me:

1. There MUST be an xml:base defined to which the reference's URI is relative to. Is this correct?
2. Where should the xml:base be defined? In the `<edmx:Edmx>` tag? For example:
```
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx" xml:base="https://www.sapdemo.com/rest/v1">
    <edmx:Reference Uri="./CORE_AP242/CsdlSchema.xml">
        <edmx:Include Alias="CORE_AP242" Namespace="CORE_APA242_SCHEMA_V_1_0"/>
    </edmx:Reference>
    ....
    ....
</edmx:Edmx>
```
3. When using this library to parse/read the OData CSDL document is there any way to get the base URL in case of a reference being used, or the full URL of the reference?

**Comment by habbes** [comment]
If you have an `IEdmModel` instance, you can use the [`IEdmModel.ReferencedModels`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.odata.edm.iedmmodel.referencedmodels) property to get a collection of the referenced models.
You can also get the list of references using the [`GetEdmReferences()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.odata.edm.extensionmethods.getedmreferences) extension method. This returns a collection of `IEdmReference` instances. With an `IEdmReference` you can get the [`Uri`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.odata.edm.iedmreference.uri) of the reference.

When parsing a CSDL document using the library's `CsdlReader`, you can specify a function to handle loading of referenced models:

```#
CsdlReader.TryParse(xmlReader, (uri) => XmlReader.Create(getStreamFromUri(uri)), ...);
```

I'm not sure how this plays out with the xml:base attribute, I don't think it's required. But it's worth a try.
